src/libsyntax/tokenstream.rs

```diff
@@ -331,35 +329,65 @@ impl TokenTree {
 /// struct itself shouldn't be directly manipulated; the internal structure is not stable,
 /// and may be changed at any time in the future. The operators will not, however (except
 /// for signatures, later on).
-#[derive(Eq,Clone,Hash,RustcEncodable,RustcDecodable)]
-pub struct TokenStream {
-    pub span: Span,
-    pub tts: Vec<TokenTree>,
+
+// #[derive(Eq,Clone,Hash,RustcEncodable,RustcDecodable)]
+#[derive(Eq,Hash,RustcEncodable,RustcDecodable)]
+pub enum TokenStream {
+  Empty,
+  Leaf    { tts : Rc<Vec<TokenTree>>, sp: Span },
+  SubLeaf { tts : Rc<Vec<TokenTree>>, offset: usize, len: usize, sp: Span },
+  Node    { left: Rc<TokenStream>, right: Rc<TokenStream>, len: usize, sp: Span},
+}
+
+
+impl clone::Clone for TokenStream {
+    fn clone(&self) -> Self {
+      match *self {
+        TokenStream::Empty => TokenStream::Empty,
+        TokenStream::Leaf { ref tts, sp} => TokenStream::Leaf { tts: tts.clone(), sp: sp },
+        TokenStream::SubLeaf { ref tts, offset, len, sp } =>
+          TokenStream::SubLeaf { tts: tts.clone(), offset: offset, len: len, sp: sp },
+        TokenStream::Node { ref left, ref right, len, sp } =>
+          TokenStream::Node { left: left.clone(), right: right.clone(), len: len, sp: sp },
+      }
+    }
+
+    // fn clone_from(&mut self, source: &Self) { ... }
 }
 
 impl fmt::Debug for TokenStream {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        if self.tts.len() == 0 {
-            write!(f, "([empty")?;
-        } else {
-            write!(f, "([")?;
-            write!(f, "{:?}", self.tts[0])?;
-
-            for tt in self.tts.iter().skip(1) {
-                write!(f, ",{:?}", tt)?;
-            }
+      match *self {
+        TokenStream::Empty => { write!(f, "") }
+        TokenStream::Leaf{ ref tts, .. } => {
+          for t in tts.iter() {
+            write!(f, "{:?}", t)?;
```

Comment:
Can you use `try!` instead of `?` so the syntex crate doesn't need to change this to compile on stable?